File: typescript/packages/extensions/src/sign-in-with-x/solana.ts
/**
 * Solana Sign-In-With-X (SIWS) support
 *
 * Implements CAIP-122 compliant message format and Ed25519 signature verification
 * for Solana wallets.
 */

import { base58 } from "@scure/base";
import nacl from "tweetnacl";
import type { CompleteSIWxInfo } from "./client";
import type { SIWxSigner } from "./sign";

/**
 * Common Solana network CAIP-2 identifiers.
 * Uses genesis hash as the chain reference per CAIP-30.
 */
export const SOLANA_MAINNET = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp";
export const SOLANA_DEVNET = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1";
export const SOLANA_TESTNET = "solana:4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z";

/**
 * Extract chain reference from CAIP-2 Solana chainId.
 *
 * @param chainId - CAIP-2 format chain ID (e.g., "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp")
 * @returns Chain reference (genesis hash)
 *
 * @example
 * ```typescript
 * extractSolanaChainReference("solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp") // "5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp"
 * ```
 */
export function extractSolanaChainReference(chainId: string): string {
  const [, reference] = chainId.split(":");
  return reference;
}

/**
 * Format SIWS message following CAIP-122 ABNF specification.
 *
 * The message format is identical to SIWE (EIP-4361) but uses "Solana account"
 * instead of "Ethereum account" in the header line.
 *
 * @param info - Server-provided extension info
 * @param address - Client's Solana wallet address (Base58 encoded public key)
 * @returns Message string ready for signing
 *
 * @example
 * ```typescript
 * const message = formatSIWSMessage(serverInfo, "BSmWDgE9ex6dZYbiTsJGcwMEgFp8q4aWh92hdErQPeVW");
 * // Returns:
 * // "api.example.com wants you to sign in with your Solana account:
 * // BSmWDgE9ex6dZYbiTsJGcwMEgFp8q4aWh92hdErQPeVW
 * //
 * // Sign in to access your content
 * //
 * // URI: https://api.example.com/data
 * // Version: 1
 * // Chain ID: mainnet
 * // Nonce: abc123
 * // Issued At: 2024-01-01T00:00:00.000Z"
 * ```
 */
export function formatSIWSMessage(info: CompleteSIWxInfo, address: string): string {
  const lines: string[] = [
    `${info.domain} wants you to sign in with your Solana account:`,
    address,
    "",
  ];

  // Statement (optional, with blank line after)
  if (info.statement) {
    lines.push(info.statement, "");
  }

  // Required fields
  lines.push(
    `URI: ${info.uri}`,
    `Version: ${info.version}`,
    `Chain ID: ${extractSolanaChainReference(info.chainId)}`,
    `Nonce: ${info.nonce}`,
    `Issued At: ${info.issuedAt}`,
  );

  // Optional fields
  if (info.expirationTime) {
    lines.push(`Expiration Time: ${info.expirationTime}`);
  }
  if (info.notBefore) {
    lines.push(`Not Before: ${info.notBefore}`);
  }
  if (info.requestId) {
    lines.push(`Request ID: ${info.requestId}`);
  }

  // Resources (optional)
  if (info.resources && info.resources.length > 0) {
    lines.push("Resources:");
    for (const resource of info.resources) {
      lines.push(`- ${resource}`);
    }
  }

  return lines.join("\n");
}

/**
 * Verify Ed25519 signature for SIWS.
 *
 * @param message - The SIWS message that was signed
 * @param signature - Ed25519 signature bytes
 * @param publicKey - Solana public key bytes (32 bytes)
 * @returns true if signature is valid
 *
 * @example
 * ```typescript
 * const messageBytes = new TextEncoder().encode(message);
 * const valid = verifySolanaSignature(message, signatureBytes, publicKeyBytes);
 * ```
 */
export function verifySolanaSignature(
  message: string,
  signature: Uint8Array,
  publicKey: Uint8Array,
): boolean {
  const messageBytes = new TextEncoder().encode(message);
  return nacl.sign.detached.verify(messageBytes, signature, publicKey);
}

/**
 * Decode Base58 string to bytes.
 *
 * Solana uses Base58 encoding (Bitcoin alphabet) for addresses and signatures.
 *
 * @param encoded - Base58 encoded string
 * @returns Decoded bytes
 * @throws Error if string contains invalid Base58 characters
 *
 * @example
 * ```typescript
 * const publicKeyBytes = decodeBase58("BSmWDgE9ex6dZYbiTsJGcwMEgFp8q4aWh92hdErQPeVW");
 * // Returns Uint8Array of 32 bytes
 * ```
 */
export function decodeBase58(encoded: string): Uint8Array {
  return base58.decode(encoded);
}

/**
 * Encode bytes to Base58 string.
 *
 * @param bytes - Bytes to encode
 * @returns Base58 encoded string
 */
export function encodeBase58(bytes: Uint8Array): string {
  return base58.encode(bytes);
}

/**
 * Detect if a signer is Solana-compatible.
 * Checks for Solana-specific properties that don't exist on EVM signers.
 *
 * @param signer - The signer to check
 * @returns true if the signer is a Solana signer
 */
export function isSolanaSigner(signer: SIWxSigner): boolean {
  // SolanaKitSigner has signMessages method (plural) - only Solana has this
  if ("signMessages" in signer && typeof signer.signMessages === "function") {
    return true;
  }
  // WalletAdapterSigner has publicKey property
  // But viem also has publicKey (as hex string), so we need to distinguish:
  // - Solana wallet adapter: publicKey has toBase58() method OR is a non-hex string
  // - viem: publicKey is a hex string starting with 0x
  if ("publicKey" in signer && signer.publicKey) {
    const pk = signer.publicKey;
    // Check for toBase58 method (Solana PublicKey object)
    if (typeof pk === "object" && pk !== null && "toBase58" in pk) {
      return true;
    }
    // Check for non-hex string (Solana base58 address)
    if (typeof pk === "string" && !pk.startsWith("0x")) {
      return true;
    }
  }
  return false;
}

File: python/x402/extensions/sign_in_with_x/solana.py
"""Solana Sign-In-With-X (SIWS) support."""

from __future__ import annotations

from typing import Any

SOLANA_MAINNET = "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp"
SOLANA_DEVNET = "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1"
SOLANA_TESTNET = "solana:4uhcVJyU9pJkvQyS88uRDiswHXSCkY3z"

_BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"


def extract_solana_chain_reference(chain_id: str) -> str:
    """Extract chain reference from CAIP-2 Solana chainId."""
    return chain_id.split(":", 1)[1]


def decode_base58(encoded: str) -> bytes:
    """Decode Base58 string to bytes."""
    if not encoded:
        return b""
    leading_zeros = len(encoded) - len(encoded.lstrip("1"))
    num = 0
    for char in encoded:
        try:
            num = num * 58 + _BASE58_ALPHABET.index(char)
        except ValueError as e:
            raise ValueError(f"Invalid Base58 character: {char}") from e
    combined = num.to_bytes((num.bit_length() + 7) // 8, "big") if num else b""
    return b"\x00" * leading_zeros + combined


def encode_base58(data: bytes) -> str:
    """Encode bytes to Base58 string."""
    if not data:
        return ""
    leading_zeros = len(data) - len(data.lstrip(b"\x00"))
    num = int.from_bytes(data, "big")
    encoded = ""
    while num > 0:
        num, rem = divmod(num, 58)
        encoded = _BASE58_ALPHABET[rem] + encoded
    return "1" * leading_zeros + (encoded or "1")


def format_siws_message(info: Any, address: str) -> str:
    """Format SIWS message following CAIP-122 ABNF."""

    def _get(name: str, alias: str | None = None) -> Any:
        if hasattr(info, name):
            return getattr(info, name)
        return info.get(alias or name)

    lines = [
        f"{_get('domain')} wants you to sign in with your Solana account:",
        address,
        "",
    ]
    statement = _get("statement")
    if statement:
        lines.extend([statement, ""])
    lines.extend(
        [
            f"URI: {_get('uri')}",
            f"Version: {_get('version')}",
            f"Chain ID: {extract_solana_chain_reference(_get('chain_id', 'chainId'))}",
            f"Nonce: {_get('nonce')}",
            f"Issued At: {_get('issued_at', 'issuedAt')}",
        ]
    )
    expiration = _get("expiration_time", "expirationTime")
    if expiration:
        lines.append(f"Expiration Time: {expiration}")
    not_before = _get("not_before", "notBefore")
    if not_before:
        lines.append(f"Not Before: {not_before}")
    request_id = _get("request_id", "requestId")
    if request_id:
        lines.append(f"Request ID: {request_id}")
    resources = _get("resources")
    if resources:
        lines.append("Resources:")
        for resource in resources:
            lines.append(f"- {resource}")
    return "\n".join(lines)


def verify_solana_signature(message: str, signature: bytes, public_key: bytes) -> bool:
    """Verify Ed25519 signature for SIWS."""
    try:
        from nacl.exceptions import BadSignatureError
        from nacl.signing import VerifyKey
    except ImportError as e:
        raise ImportError(
            "SIWX Solana support requires PyNaCl. Install with: pip install x402[extensions]"
        ) from e

    try:
        VerifyKey(public_key).verify(signature + message.encode("utf-8"))
        return True
    except BadSignatureError:
        return False


def is_solana_signer(signer: Any) -> bool:
    """Detect if a signer is Solana-compatible."""
    if hasattr(signer, "sign_messages") and callable(getattr(signer, "sign_messages", None)):
        return True
    # x402 KeypairSigner wraps solders.Keypair
    if hasattr(signer, "keypair"):
        return True
    public_key = getattr(signer, "public_key", None) or getattr(signer, "publicKey", None)
    if public_key is not None:
        if hasattr(public_key, "to_base58") or hasattr(public_key, "toBase58"):
            return True
        if isinstance(public_key, str) and not public_key.startswith("0x"):
            return True
    return False
